### PR TITLE
Deathmatch gamemode

### LINKED ISSUE #691: [FEATURE] Deathmatch
### Description

A deathmatch mode where there's no team, players spawn at non-team spawns, loadouts pick on spawn, etc...

### Proposed Solution

- [x] Deathmatch spawning determined by external file or map's DM spawn entity
- [x] Respawning (TDM + idle already have that)
- [x] Loadout picks when spawning in
- [x] Can be turned on similar to TDM and VIP for the time being
- [x] No team for deathmatch mode, refit the source code to handle both team and non-teamplay
- [x] Either to EX: 50 kills or timer runs out
- [ ] Scoreboard change to deal with no teams

### Alternatives Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by nullsystem**
## Description
* **TODO: Weapon restriction? (Separate PR)**
* Deathmatch spawning
    * Use info_player_deathmatch entity if possible
    * Otherwise can just use the separate file (dmspawn)
    * Vector of spawns checked if it been filled at the current fill, spawned in starting from a given random int to the next available spawn
* Deathmatch gameplay
    * Either to maximum XP or time runs out
    * If tying at max XP or post time out, it'll end when there's a single highest XP winner
* UI
    * HUD round state adjusted for DM
    * Scoreboard adjusted for DM, no Jinrai/NSF coloring
* ConVars
    * neo_sv_dmspawn_useent - Default 0, override to use map's deathmatch spawn entity over the dmspawn loc file
    * neo_sv_dm_max_class_dur - Default 5s, maximum duration you're allowed to pick a class post-spawning in
    * neo_dm_round_timelimit - Timelimit for deathmatch
    * neo_sv_dm_win_xp - Default 50, if 0 disable XP win limit
* ConCommands - Here's the commands to edit dmspawn loc files and check on it. Used only for local server.
    * neo_sv_dmspawn_create - Create a dmspawn spawn
    * neo_sv_dmspawn_removeall - Remove all dmspawn spawns
    * neo_sv_dmspawn_printlocs - Print all coordinates of the dmspawn spawns
    * neo_sv_dmspawn_save - Save the map's loc dmspawn file
    * neo_sv_dmspawn_load - Load the map's loc dmspawn file
    * neo_sv_dmspawn_teleportnext - Teleport to the next dmspawn spawn
    * neo_sv_dmspawn_mapinfo - Check amount of dmspawn spawns and if it can DM

## Toolchain

- Linux GCC Distro Native Arch/GCC 14

## Linked Issues

* fixes #691

### CODE REVIEW

File: mp/src/game/server/neo/neo_player.cpp [hunk 1]
```diff
@@ -1663,7 +1663,8 @@ void CNEO_Player::SetPlayerCorpseModel(int type)
 
 float CNEO_Player::GetReceivedDamageScale(CBaseEntity* pAttacker)
 {
-	if (NEORules()->GetGameType() == NEO_GAME_TYPE_TDM && m_aliveTimer.IsLessThen(5.f))
+	if ((NEORules()->GetGameType() == NEO_GAME_TYPE_TDM || NEORules()->GetGameType() == NEO_GAME_TYPE_DM)
+			&& m_aliveTimer.IsLessThen(5.f))
```
Comment (brysondev) [inline mp/src/game/server/neo/neo_player.cpp:1667]: Is this persistent after the user fires? That and 5 seconds is generous 🤔

**Review by brysondev — CHANGES_REQUESTED** [review]

**Review by brysondev — APPROVED** [review]

**Comment by AdamTadeusz** [comment]
Would it be worth giving players a few seconds of invulnerability after spawning?

**Comment by AdamTadeusz** [comment]
> > Would it be worth giving players a few seconds of invulnerability after spawning?
> 
> https://github.com/NeotokyoRebuild/neo/pull/694/files#diff-c4ea17be8fcce45e76e46a0f050cbdc12308a47653133f08ee36d7fbdf9e181dR1667
> 
> He is, but it's only until they fire or the timer expires. Or do you mean it's not worth giving invuln?

my bad, bots start shooting instantly by default so i didnt notice

**Comment by AdamTadeusz** [comment]
I do think it would be worth if the spawns saved the player's look direction though. If you're placing a spawn in a corner you probably don't want the player on spawn to face into a wall

**Comment by AdamTadeusz** [comment]
@nullsystem do you have an example map with player_info_deathmatch entities to test whether spawning on those works too?

**Comment by nullsystem** [comment]
> @nullsystem do you have an example map with player_info_deathmatch entities to test whether spawning on those works too?
> 
> [Edit] nvm works on dm_lockdown

I do plan on releasing zigzag with dm entity spawn soon anyway and probably a dmspawn file for oilstain, both for playtesting this normally

**Comment by AdamTadeusz** [comment]
https://github.com/user-attachments/assets/3145b5ab-5836-49db-812a-7c86f1b4f426

Shame we decoupled both aux and sprint from these, maybe the energy packs could refill your thermoptic cloak

**Review by AdamTadeusz — DISMISSED** [review]

**Comment by nullsystem** [comment]
@AdamTadeusz @brysondev I've added in y-look-angle for each DMSpawn spawn now

**Review by brysondev — APPROVED** [review]

**Review by AdamTadeusz — APPROVED** [review]

**Pull request merged**